### PR TITLE
[FLINK-20616][python] Support row-based operation to accept user-defined function directly

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will support row-based operation to accept user-defined function directly*

## Verifying this change

This change added tests and can be verified as follows:

  - *IT test in `test_row_based_operation.py`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/pyflink/table/table.py [hunk 1]
```diff
@@ -1059,9 +1087,26 @@ def flat_aggregate(self, func: Union[str, Expression]) -> 'FlatAggregateTable':
...
         if isinstance(func, str):
             return FlatAggregateTable(self._j_table.flatAggregate(func), self._t_env)
+        elif isinstance(func, Expression):
+            return FlatAggregateTable(self._j_table.flatAggregate(func._j_expr), self._t_env)
         else:
+            func = self._without_columns(func)
             return FlatAggregateTable(self._j_table.flatAggregate(func._j_expr), self._t_env)
 
+    def _without_columns(self, func: UserDefinedAggregateFunctionWrapper) -> Expression:
+        get_gateway()
```
Comment (dianfu) [inline flink-python/pyflink/table/table.py:1097]: could be removed?

File: flink-python/pyflink/table/table.py [hunk 2]
```diff
@@ -1059,9 +1087,26 @@ def flat_aggregate(self, func: Union[str, Expression]) -> 'FlatAggregateTable':
...
+        group_keys_field = self._j_table.getClass().getDeclaredField("groupKeys")
+        group_keys_field.setAccessible(True)
+        j_group_keys = group_keys_field.get(self._j_table)
+        without_keys = without_columns(
```
Comment (dianfu) [inline flink-python/pyflink/table/table.py:1101]:
```suggestion
        fields_without_keys = without_columns(
```

File: flink-python/pyflink/table/table.py [hunk 3]
```diff
@@ -1165,9 +1211,30 @@ def aggregate(self, func: Union[str, Expression]) -> 'AggregatedTable':
...
         if isinstance(func, str):
             return AggregatedTable(self._j_table.aggregate(func), self._t_env)
+        elif isinstance(func, Expression):
+            return AggregatedTable(self._j_table.aggregate(func._j_expr), self._t_env)
         else:
+            func = self._without_columns(func)
             return AggregatedTable(self._j_table.aggregate(func._j_expr), self._t_env)
 
+    def _without_columns(self, func: UserDefinedAggregateFunctionWrapper) -> Expression:
+        get_gateway()
```
Comment (dianfu) [inline flink-python/pyflink/table/table.py:1221]: could be removed?

File: flink-python/pyflink/table/table.py [hunk 4]
```diff
@@ -1165,9 +1211,30 @@ def aggregate(self, func: Union[str, Expression]) -> 'AggregatedTable':
         """
         if isinstance(func, str):
             return AggregatedTable(self._j_table.aggregate(func), self._t_env)
+        elif isinstance(func, Expression):
+            return AggregatedTable(self._j_table.aggregate(func._j_expr), self._t_env)
         else:
+            func = self._without_columns(func)
             return AggregatedTable(self._j_table.aggregate(func._j_expr), self._t_env)
 
+    def _without_columns(self, func: UserDefinedAggregateFunctionWrapper) -> Expression:
```
Comment (dianfu) [inline flink-python/pyflink/table/table.py:1220]:
```suggestion
    def _to_expr(self, func: UserDefinedAggregateFunctionWrapper) -> Expression:
```

File: flink-python/pyflink/table/tests/test_row_based_operation.py [hunk 5]
```diff
@@ -125,9 +123,11 @@ def test_aggregate_with_pandas_udaf(self):
                                [DataTypes.FIELD("a", DataTypes.FLOAT()),
                                 DataTypes.FIELD("b", DataTypes.INT())]),
                            func_type="pandas")
-        t.group_by(t.a) \
-            .aggregate(pandas_udaf(t.b).alias("c", "d")) \
-            .select("a, c, d").execute_insert("Results") \
+        t.select(t.a, t.b) \
+            .group_by(t.a) \
+            .aggregate(pandas_udaf) \
+            .select("*") \
```
Comment (dianfu) [inline flink-python/pyflink/table/tests/test_row_based_operation.py:129]: I guess `.select("*")`  could be removed?

File: flink-python/pyflink/table/table.py [hunk 6]
```diff
@@ -759,7 +761,7 @@ def drop_columns(self, *fields: Union[str, Expression]) -> 'Table':
             assert isinstance(fields[0], str)
             return Table(self._j_table.dropColumns(fields[0]), self._t_env)
 
-    def map(self, func: Union[str, Expression]) -> 'Table':
+    def map(self, func: Union[str, Expression, UserDefinedScalarFunctionWrapper]) -> 'Table':
```
[the remark this answers is not in the document]
Comment (HuangXingBo) [reply flink-python/pyflink/table/table.py:764]: Yes. make sense.

File: flink-python/pyflink/table/tests/test_row_based_operation.py (hunk 5 above)
Comment (HuangXingBo) [reply flink-python/pyflink/table/tests/test_row_based_operation.py:129]: `AggregateTable` need to call `select` method to convert to `Table`

File: flink-python/pyflink/table/table.py [hunk 7]
```diff
@@ -1059,9 +1098,25 @@ def flat_aggregate(self, func: Union[str, Expression]) -> 'FlatAggregateTable':
...
 
+    def _to_expr(self, func: UserDefinedAggregateFunctionWrapper) -> Expression:
+        group_keys_field = self._j_table.getClass().getDeclaredField("groupKeys")
+        group_keys_field.setAccessible(True)
+        j_group_keys = group_keys_field.get(self._j_table)
+        fields_without_keys = without_columns(
+            j_group_keys[0], *([j_group_keys[i] for i in range(1, len(j_group_keys))]))
+        if hasattr(func, "alias_names"):
+            alias_names = getattr(func, "alias_names")
+            func_expression = func(fields_without_keys).alias(*alias_names)
```
Comment (dianfu) [inline flink-python/pyflink/table/table.py:1115]: I think the keys should also be given to the row based function.

File: flink-python/pyflink/table/table.py [hunk 8]
```diff
@@ -817,6 +828,8 @@ def aggregate(self, func: Union[str, Expression]) -> 'AggregatedTable':
             ...                    DataTypes.FIELD("b", DataTypes.INT())]),
             ...               func_type="pandas")
             >>> tab.aggregate(agg(tab.a).alias("a", "b")).select("a, b")
+            >>> # input all columns
+            >>> tab.aggregate(agg.alias("a, b")).select("a, b")
```
Comment (dianfu) [inline flink-python/pyflink/table/table.py:832]: For Row based operation, the row function should receive Row/DataFrame as input. Otherwise, the fields name information will be missing. This will become a problem when there are hundreds of columns. It may be difficult to use as users have to give each column a meaningful name.

File: flink-python/pyflink/table/table.py [hunk 9]
```diff
@@ -792,6 +798,8 @@ def flat_map(self, func: Union[str, Expression]) -> 'Table':
             ...     for s in string.split(","):
             ...         yield x, s
             >>> tab.flat_map(split(tab.a, table.b))
+            >>> # input all columns
```
Comment (dianfu) [inline flink-python/pyflink/table/table.py:801]:
```suggestion
            >>> # take all the columns as inputs
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/python/PythonTableFunction.java [hunk 10]
```diff
@@ -48,6 +48,7 @@
 	private final PythonFunctionKind pythonFunctionKind;
 	private final boolean deterministic;
 	private final PythonEnv pythonEnv;
+	private final boolean usedInRowBasedOperation;
```
Comment (dianfu) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/python/PythonTableFunction.java:51]: I guess we could also add row-based support in join_lateral/left_outer_join_lateral? I'm fine to do it in a separate PR if it you want.

File: flink-python/pyflink/fn_execution/operation_utils.py [hunk 11]
```diff
@@ -43,6 +43,15 @@ def wrap_pandas_result(it):
     return arrays
 
 
+def wrap_inputs_to_row(*args):
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/operation_utils.py:46]:
```suggestion
def wrap_inputs_as_row(*args):
```

File: flink-python/pyflink/table/tests/test_row_based_operation.py [hunk 12]
```diff
@@ -120,14 +117,16 @@ def test_aggregate_with_pandas_udaf(self):
...
                            result_type=DataTypes.ROW(
                                [DataTypes.FIELD("a", DataTypes.FLOAT()),
                                 DataTypes.FIELD("b", DataTypes.INT())]),
                            func_type="pandas")
-        t.group_by(t.a) \
-            .aggregate(pandas_udaf(t.b).alias("c", "d")) \
-            .select("a, c, d").execute_insert("Results") \
+        t.select(t.a, t.b) \
+            .group_by(t.a) \
+            .aggregate(pandas_udaf) \
```
Comment (dianfu) [inline flink-python/pyflink/table/tests/test_row_based_operation.py:127]: could you improve pandas_udaf to let it also access the group key?

**Review by dianfu — COMMENTED** [review]
@HuangXingBo Thanks a lot for the update.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/python/PythonTableFunction.java (hunk 10 above)
Comment (HuangXingBo) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/python/PythonTableFunction.java:51]: Yes. I have created https://issues.apache.org/jira/browse/FLINK-20712 to support this feature.

**Comment by HuangXingBo** [comment]
@dianfu Thanks a lot for the review. I have addressed the comments at the latest commit.

**Review by dianfu — APPROVED** [review]
